src/test/java/reactor/netty/TomcatServer.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2011-2019 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.netty;
+
+import org.apache.catalina.Context;
+import org.apache.catalina.Wrapper;
+import org.apache.catalina.startup.Tomcat;
+
+import javax.servlet.MultipartConfigElement;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.Part;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Collection;
+
+/**
+ * @author Violeta Georgieva
+ */
+public class TomcatServer {
+	static final String TOMCAT_BASE_DIR = "./build/tomcat";
+
+	final Tomcat tomcat;
+
+	boolean started = false;
+
+	public TomcatServer() {
+		this(0);
+	}
+
+	public TomcatServer(int port) {
+		this.tomcat = new Tomcat();
+		this.tomcat.setPort(port);
+		File baseDir = new File(TOMCAT_BASE_DIR);
+		this.tomcat.setBaseDir(baseDir.getAbsolutePath());
+	}
+
+	public int port() {
+		if (this.started) {
+			return this.tomcat.getConnector().getLocalPort();
+		} else {
+			throw new IllegalStateException("Tomcat is not started.");
+		}
+	}
+
+	public void start() throws Exception {
+		if (!this.started) {
+			this.tomcat.start();
+			this.started = true;
+		}
+	}
+
+	public void stop() throws Exception {
+		if (this.started) {
+			this.tomcat.stop();
+			this.tomcat.destroy();
+			this.started = false;
+		}
+	}
+
+	public void createDefaultContext() {
```

Comment:
perhaps the `createDefaultContext()` call could be made in `start()` so that an extra call is avoided in each `HttpClientTest` that make use of the `TomcatServer`?

Have you tried to start the server once in a `@ClassRule` btw? this would at least avoid the ceremony around starting and stopping the server (see below).